File: Sample_Study/ETL_Study.java
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ELT_Study {
    public static void main(String[] args) throws Exception {
        String sourceUrl = "jdbc:postgresql://source_host/db";
        String targetUrl = "jdbc:postgresql://target_host/dw";
        
        List<Order> extractedData = new ArrayList<>();

        // 1. Extract: 소스 DB에서 데이터 읽기
        try (Connection srcConn = DriverManager.getConnection(sourceUrl, "user", "pass");
             Statement stmt = srcConn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT id, name, amount FROM orders")) {
            
            while (rs.next()) {
                extractedData.add(new Order(
                    rs.getInt("id"),
                    rs.getString("name"),
                    rs.getDouble("amount")
                ));
            }
        }

        // 2. Transform: 자바 애플리케이션 내에서 데이터 변환 (소문자 -> 대문자)
        for (Order order : extractedData) {
            order.setName(order.getName().toUpperCase());
        }

        // 3. Load: 변환된 데이터를 타겟 데이터 웨어하우스에 적재
        try (Connection tgtConn = DriverManager.getConnection(targetUrl, "user", "pass");
             PreparedStatement pstmt = tgtConn.prepareStatement("INSERT INTO fact_orders (id, name, amount) VALUES (?, ?, ?)")) {
            
            for (Order order : extractedData) {
                pstmt.setInt(1, order.getId());
                pstmt.setString(2, order.getName());
                pstmt.setDouble(3, order.getAmount());
                pstmt.addBatch();
            }
            pstmt.executeBatch();
        }
    }
    
    // 단순 DTO 클래스
    static class Order {
        private int id; private String name; private double amount;
        public Order(int id, String name, double amount) { this.id = id; this.name = name; this.amount = amount; }
        public int getId() { return id; }
        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        public double getAmount() { return amount; }
    }
}

File: Sample_Study/ETL_Study.py


import pandas as pd
from sqlalchemy import create_engine

# 1. Extract: 소스 DB 또는 파일에서 원시 데이터 추출
source_engine = create_engine('postgresql://user:pass@source_host/db')
df = pd.read_sql("SELECT id, name, amount FROM orders", source_engine)

# 2. Transform: 파이썬(Pandas) 메모리 내에서 데이터 정제 및 변환
# 이름을 대문자로 변경하는 작업 수행
df['name'] = df['name'].str.upper()

# 3. Load: 타겟 데이터 웨어하우스에 정제된 데이터 적재
target_engine = create_engine('postgresql://user:pass@target_host/dw')
df.to_sql('fact_orders', target_engine, if_exists='append', index=False)